File: GregBot.Modules.Parrot/ParrotModule.fs
namespace GregBot.Modules.Parrot

open System.Threading.Tasks
open GregBot.Domain.Attributes
open GregBot.Domain.Events
open GregBot.Domain
open GregBot.Domain.Interfaces
open GregBot.Modules.Parrot.Matchers
open Microsoft.Extensions.Logging

[<Priority(Priority.Low)>]
type ParrotModule(logger: ILogger<ParrotModule>, messageService: IMessageService) =
    let ReplyFor = Seq.reduce (Combine Option.orElse) Rules.All
    
    [<Priority(Priority.High)>]
    let OnMessage (event: MessageEvent) =
        let message = event.Message
        
        if messageService.IsSentBySelf message
        then Task.CompletedTask else
            let content = message.Content
            
            match ReplyFor content with
            | None -> Task.CompletedTask
            | Some(reply) ->
                logger.LogDebug("Replied with \'{Reply}\' to the message \'{Message}\'", reply, content)
                messageService.Reply(reply, message)
    
    member _.Name = "Parrot"
    member _.Activate (bot: GregBot) =
        bot.OnMessage.Subscribe(OnMessage)
    
    interface IModule with        
        member this.Name = this.Name
        member this.Activate bot = this.Activate bot
    

File: Gregbot.Modules.Parrot/ParrotModule.cs
using GregBot.Domain;
using GregBot.Domain.Attributes;
using GregBot.Domain.Events;
using GregBot.Domain.Interfaces;
using GregBot.Domain.Services;
using Microsoft.Extensions.Logging;

namespace GregBot.Modules.Parrot;

[Priority(Priority.Low)]
public class ParrotModule : IModule
{
    private readonly ILogger<ParrotModule> _logger;
    private readonly IEnumerable<ParrotRule> _rules;
    private readonly IMessageService _messageService;

    public ParrotModule(ParrotRuleProvider ruleProvider, ILogger<ParrotModule> logger, IMessageService messageService)
    {
        _rules = ruleProvider();
        _logger = logger;
        _messageService = messageService;
    }
    
    public string Name => "Parrot";
    public void Activate(IGregBot bot)
    {
        bot.OnMessage.Subscribe(OnMessage);
    }

    [Priority(Priority.High)]
    private Task OnMessage(MessageEvent @event)
    {
        var message = @event.Message;

        if (_messageService.IsSentBySelf(message))
            return Task.CompletedTask;

        var content = message.Content;
        foreach (var rule in _rules)
        {
            var reply = rule(content);
            if (reply is null) continue;
            
            _logger.LogDebug("Replied with '{Reply}' to '{Message}'", reply, content);
            return _messageService.ReplyTo(message, reply);
        }

        return Task.CompletedTask;
    }
}